### ISSUE TITLE
Package-install fails

### ISSUE THREAD

**Issue by mjayuk** [hlissner/doom-emacs#258]
A recent convert from spacemacs (and therefore a relative newbie to DOOM) - I have one or perhaps two separate problems:

Firstly, when I try and use swiper (I use helm) I receive the following error:

_Wrong type argument: commandp, swiper_

Thinking that I could try and install/reinstall swiper for helm (I couldn't find and swiper.el files in my .emacs.d folder) I tried to install the swiper-helm package.  This didn't work and I received the following error:

_helm-M-x: Certificate validation failed melpa.org, verification code 142_

I seem to receive this error when I try and install any package from within emacs.

I’ve tried running Make Doctor - which doesn't report any problems.  I've separately run gnutls-cli melpa.org in terminal and connected ok.

Any suggestions, help would be appreciated

**Comment by hlissner** [comment]
> Wrong type argument: commandp, swiper

This may indicate an outdated autoloads file. Try `make autoloads` and see if that resolves this issue. I recommend running `make` any time you edit `init.el` (it is equivalent to `make autoremove install autoloads`, which will auto-remove unneeded plugins, install missing ones, and regenerate the autoloads file).

> helm-M-x: Certificate validation failed melpa.org, verification code 142

Curious. I assume you ran `make install` before you ran Doom the first time and it was successful? If so, that (and doom doctor's lack of complaints) _should_ indicate that TLS validation is working fine for you.

Is it possible this was a one time thing? Have you tried again since? I do get validation errors with melpa about once a month (I don't have that issue with elpa). Occurrences are too seldom to reliably test for.

**Comment by mjayuk** [comment]
Hi 

Many thanks for your response and help.

I had come back top the site today to report that:

1. I switched Helm to Ivy and this stopped the commandp errors, swiper seemed to be working again.  Changing back to Helm caused commandp error to reappear.  For information, each time after changing the init.el file I had run make install.  I did notice when I changed from Ivy to Helm that swiper was removed, and when changing back to Ivy swiper was reinstalled.

Q. Do you have any instruction on when to run Make Autoloads, Make etc.?

2. This one is now working again and it's difficult to pin down exactly what happened.  I'm pretty sure that it happened after running ' Make Autoloads'.  You asked whether this could have been a one time thing.  The problem affected me for a number of days on different machines (all with the same configuration I'd add).  Again, since running Make Autoloads the problem seems to have disappeared.

Finally and I appreciate that you may want me to open a second ticket for this question, but are you aware of a fix that would allow someone using a GB keyboard type the hash (#) key.  There are a number of emacs and Sublime solutions posted online, but I’ve not managed to get one of them to work.

**Comment by mjayuk** [comment]
Err... for Sublime please read Spacemacs,  Sorry about that, I had sublime open in the background and must have subconsciously thought of that instead.

**Comment by hlissner** [comment]
Sorry for the late response!

> Q. Do you have any instruction on when to run Make Autoloads, Make etc.?

`make` is the only command you need to worry about, when using doom. You should run it every time you edit `~/.emacs.d/init.el` (more specifically, the `doom!` block within). 

However, it is helpful to understand what `make autoloads` does: it generates a special file that tells Emacs where to find lazy-loaded functions and commands. Without it, you get commandp, void-variable and void-function errors because Emacs doesn't know where to look.

> I switched Helm to Ivy and this stopped the commandp errors

This happens because the first time you ran `make install`, after downloading Doom, it auto-generated your autoloads file (essentially, it ran `make autoloads` _for you_ the first time). Since Ivy is the default completion engine in Doom -- and was likely enabled at the beginning -- the autoloads file included ivy's commands. That is why ivy ran flawlessly.

> I did notice when I changed from Ivy to Helm that swiper was removed, and when changing back to Ivy swiper was reinstalled.

The problem here is that only the ivy module "asks for" swiper. The helm module does not, so Doom thinks it's an unneeded package when you switch back to helm. In order to make the package "permanent", you need to declare it in a `packages.el` file, possibly in your private module.

tl;dr If you want to use helm _and_ swiper, add the following to your private module (in `modules/private/$USER`):

```emacs-lisp
;; in modules/private/$USER/packages.el
(package! swiper)
(package! swiper-helm)

;; any other packages you want here
```

```emacs-lisp
;; in modules/private/$USER/config.el
(def-package! swiper :commands (swiper swiper-all))
(def-package! swiper-helm :commands (swiper-helm))

;; other customizations here
```

Then run `make` and you should be golden.

> are you aware of a fix that would allow someone using a GB keyboard type the hash (#) key. There are a number of emacs and Sublime solutions posted online, but I’ve not managed to get one of them to work.

I'm afraid I don't fully understand the question. Please link me one of those solutions you mentioned, that may help.

**Comment by hlissner** [comment]
Oh, I should mention that the helm module already has a plugin that is similar to swiper called [helm-swoop](https://github.com/ShingoFukuyama/helm-swoop) (`M-x helm-swoop`). Perhaps that would suit you better? It could save you the hassle if so.

**Comment by hlissner** [comment]
Closing this due to inactivity. Feel free to reopen if your issue hasn't been resolved.
